MaxText/decode.py

```diff
@@ -134,14 +134,14 @@ def decode_loop(config, state=None):
   # Model and Optimizer definition
   model = Transformer(config, mesh = mesh)
 
-  tx = optax.adamw(
-    max_utils.create_learning_rate_schedule(config)
-  ) # TODO: we need an optax.GradientTransformation to form a TrainState, but we don't use it when decoding
-
-
   _, sp_tokenizer = create_data_iterator_with_tokenizer(config, mesh)
 
-  state, state_mesh_annotations = max_utils.setup_initial_state(model, tx, config, rng, mesh, checkpoint_manager)
+  state, state_mesh_annotations = max_utils.setup_initial_state(
+    model, None, config, rng, mesh, checkpoint_manager, True
+    )
+  assert state.opt_state == {}, "non null opt_state in checkpoint"
+  num_params = max_utils.calculate_num_params_from_pytree(state.params)
```

Comment:
I think it is weird we have a strange TrainState, that really seems not ideal. Wouldn't it be better to just not have a train state? Is that hard for some reason?

Reply:
updated functions to make it cleaner.

Reply:
yes looks amazingly clear!